src/Expr.cc

```diff
@@ -4136,15 +4136,15 @@ RecordCoerceExpr::RecordCoerceExpr(Expr* op, RecordType* r)
 
 			if ( ! same_type(sup_t_i, sub_t_i) )
 				{
-				if ( sup_t_i->Tag() != TYPE_RECORD ||
-				     sub_t_i->Tag() != TYPE_RECORD ||
-				     ! record_promotion_compatible(sup_t_i->AsRecordType(),
-				                                   sub_t_i->AsRecordType()) )
+				if ( ! BothArithmetic(sup_t_i->Tag(), sub_t_i->Tag())  &&
```

Comment:
I originally had this code do all of the bounds checking before I started looking at `check_and_promote`. `check_and_promote` does erroneously fail if for example you try to convert a double into a count, even if the value is low enough to fit.

Reply:
Looking at this I don't think it's worth trying to narrow any further right here. This block is just checking whether the types are compatible at all. All of the numeric types are technically compatible. There could theoretically be more bounds checking in `Fold` prior to calling `check_and_promote`, or the latter could do better bounds checking itself.

Reply:
> Looking at this I don't think it's worth trying to narrow any further right here. This block is just checking whether the types are compatible at all. All of the numeric types are technically compatible.

> `check_and_promote` does erroneously fail if for example you try to convert a double into a count, even if the value is low enough to fit.

That's what I'm talking about, though: it's not great for us to treat `double` -> `count` conversion as a run-time error when we could have detected that at parse-time and prevented going any further.

And it's maybe fine to treat this as a general error for now, even for the "technically, 1.0 can safely fit into a count" cases, since it's hard to imagine many legit use-cases where one wants to do that on purpose (and for those that do, we have conversion functions like `double_to_count` which would bypass any safety checks).

Reply:
I had considered writing a `would_overflow` method that takes two types and an input value, and returns true or false if the value would overflow. That might be useful both here and in `check_and_promote`. If we knew it was going to overflow here we could just give up parsing it, and it could fix the double->count case.